yb-voyager/src/queryparser/traversal.go

```diff
@@ -0,0 +1,173 @@
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.String
+	Field: sval, Type: string
+	Field: location, Type: int32
+	Field: xmloption, Type: enum
+	Field: location, Type: int32
+	Field: location, Type: int32
+	Field: from_clause, Type: message
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.RangeVar
+	Field: relname, Type: string
+	Field: inh, Type: bool
+	Field: relpersistence, Type: string
+	Field: location, Type: int32
+	Field: limit_option, Type: enum
+	Field: op, Type: enum
+*/
+
+func TraverseParseTree(msg protoreflect.Message, visited map[protoreflect.Message]bool, processor NodeProcessor) error {
+	if msg == nil || !msg.IsValid() {
+		return nil
+	}
+
+	if visited[msg] {
+		return nil
+	}
+	visited[msg] = true
+
+	nodeType := msg.Descriptor().FullName()
+	log.Debugf("Traversing NodeType: %s\n", nodeType)
+	// applying the processor to the current node
+	if err := processor(msg); err != nil {
+		log.Debugf("error processing node %s: %v", nodeType, err)
+		return fmt.Errorf("error processing node %s: %w", nodeType, err)
+	}
+
+	// Reference Oneof - https://protobuf.dev/programming-guides/proto3/#oneof
+	if nodeType == PG_QUERY_NODE_NODE {
+		nodeField := msg.WhichOneof(msg.Descriptor().Oneofs().ByName("node"))
+		if nodeField != nil {
+			value := msg.Get(nodeField)
+			if value.IsValid() {
+				err := TraverseParseTree(value.Message(), visited, processor)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return TraverseNodeFields(msg, visited, processor)
+}
+
+func TraverseNodeFields(msg protoreflect.Message, visited map[protoreflect.Message]bool, processor NodeProcessor) error {
+	fields := msg.Descriptor().Fields()
+	if fields == nil || fields.Len() == 0 {
+		return nil
+	}
+
+	for i := 0; i < fields.Len(); i++ {
+		fieldDesc := fields.Get(i)
+
+		// check if the field is set or not (for example: fields with DEFAULT values(set) won't be NULL)
+		if !msg.Has(fieldDesc) {
+			log.Debugf("field %q is not a part of the message ", fieldDesc.FullName())
+			continue
+		}
+
+		// (this check might be redundant due to previous Has() check)
+		// Handle oneof fields
+		if oneof := fieldDesc.ContainingOneof(); oneof != nil {
+			// Only process if the fieldDesc (field) is the one that is set in the oneof group
+			if fieldDesc != msg.WhichOneof(oneof) {
+				continue // Skip unset fields in the oneof group
+			}
+		}
+		log.Debugf("Field: %s, Type: %s\n", fieldDesc.Name(), fieldDesc.Kind())
+		value := msg.Get(fieldDesc)
+		switch {
+		case fieldDesc.IsList():
+			list := value.List()
+			for j := 0; j < list.Len(); j++ {
+				elem := list.Get(j)
+				// every elem in the list have same Kind
+				if fieldDesc.Kind() == protoreflect.MessageKind {
+					err := TraverseParseTree(elem.Message(), visited, processor)
+					if err != nil {
+						log.Debugf("error traversing field %s: %v", fieldDesc.Name(), err)
+						return fmt.Errorf("error traversing field %s: %w", fieldDesc.Name(), err)
```

Comment:
Here, as well, if there is an error whlie processing some node, do we want to fail altogether? I would think we logError and move on?

Reply:
Top level, inside assessMigrationCommand.go, we won't fail entire assessment but for that particular query.

But here i think our assumption is that our traversal algo is full proof. Ideally should not fail to traverse the fields.
In case it does, i think we would need to revisit the algo

Reply:
Got it. Even in this case, you might want to ignore if traversing some particular node fails, but yeah I get your point. Seems okay